### PR TITLE
test(qwen25-3b-awq): LMCache vs no-LMCache A/B load test (Approach B …

### PR DESCRIPTION
**Pull request by benie-joy-possi**
## 1. Summary

This PR changes:

- Full experiment report on the load testing

It solves:

- #617 

---

## 2. Intent

The intent of this PR is:

> To experimentallly demonstrate the Lmcache reduces latency when GPU KV-cache eviction forces vLLM's built in APC to miss.

---

### In Scope

- artillery load testing config 
- Node.js generator scripts for reproducible CSV generation
- - Full experiment reprot covering all 4 iterations 
- Raw test outputs (JSON+HTML) for both variants

### Out of Scope

- Lmcache multiprocess mode

---

## 4. Verification

I verified this change by:

- [ ] Running automated tests
- [x] Running manual tests
- [x] Checking logs
- [x] Checking metrics
- [x] Testing error cases
- [ ] Testing permissions/security behavior
- [ ] Testing rollback or failure behavior, if relevant

Commands run:

```bash
helm apply 
```

Results:

```text
[paste result or link]
```

---

## 5. Screenshots / Evidence

Add evidence here:

image
image

---

## 6. Risk Assessment

Risk level:

* [x] Low
* [ ] Medium
* [ ] High

Potential risks:

* lmcache CPU sizing

Mitigation:

* CPU cache size (LMCACHE_MAX_LOCAL_CPU_SIZE  is a tunable parameter 

---

## 7. AI Usage Declaration

AI was used for:

* [x] Understanding existing code
* [x] Generating code
* [ ] Refactoring
* [ ] Generating tests
* [x] Drafting documentation
* [ ] Reviewing the diff
* [ ] Not used

Human verification:

* [x] I understand every meaningful change in this PR
* [ ] I checked generated code manually
* [ ] I checked generated tests manually
* [x] I removed unsupported AI assumptions
* [x] I accept responsibility for this PR

---

## 8. Reviewer Focus

Please focus your review on:

* [x] Correctness
* [ ] Architecture
* [ ] Security
* [ ] Performance
* [x] Tests
* [x] Maintainability
* [ ] Product intent
* [ ] Edge cases

### CODE REVIEW

File: plans/artillery/justfile [hunk 1]
```diff
@@ -274,6 +461,26 @@ help:
     @echo "  just lua-overhead         - Full Lua with/without test"
     @echo "  just lua-compare          - Compare existing Lua results"
     @echo ""
```
Comment (Koufan-De-King) [inline plans/artillery/justfile:276]: Let's remove the previous envoy performance benchmarks entirely.

**Review by Koufan-De-King — CHANGES_REQUESTED** [review]

File: plans/artillery/reports/experiment-2-approach-b-eviction.md [hunk 2]
```diff
@@ -0,0 +1,346 @@
...
+| Successful | **250 (100%)** | **250 (100%)** | — |
+| Timeouts | **0** | **0** | — |
+| Min latency | 132 ms | 120 ms | — |
+| Mean latency | 4,660.1 ms | 665.3 ms | **-3,994.8 ms (-85.7%)** |
+| **p50 (median)** | **4,147.4 ms** | **210.6 ms** | **-3,936.8 ms (-94.9%)** |
+| p75 | 5,711.5 ms | 788.5 ms | -4,923.0 ms (-86.2%) |
+| p90 | 10,407.3 ms | 1,022.7 ms | -9,384.6 ms (-90.2%) |
+| **p95** | **11,501.8 ms** | **2,101.1 ms** | **-9,400.7 ms (-81.7%)** |
+| **p99** | **12,711.5 ms** | **5,272.4 ms** | **-7,439.1 ms (-58.5%)** |
+| Max latency | 15,623 ms | 7,987 ms | — |
```
Comment (stephane-segning) [inline plans/artillery/reports/experiment-2-approach-b-eviction.md]: Delta here cannot be "–"

File: plans/artillery/reports/report-qwen25-huge-prefix-lmcache.html [hunk 3]
```diff
@@ -0,0 +1,511 @@
...
+            <td style="color: #22c55e">0</td>
+            <td>1,022.7 ms</td>
+            <td>1,978.7 ms</td>
+            <td>1,978.7 ms</td>
+          </tr>
+          
+          <tr>
+            <td>2</td>
+            <td>19</td>
+            <td style="color: #22c55e">20</td>
```
Comment (stephane-segning) [inline plans/artillery/reports/report-qwen25-huge-prefix-lmcache.html:251]: 19 requests and 20 success? How come?

File: plans/artillery/reports/report-qwen25-huge-prefix-lmcache.html [hunk 4]
```diff
@@ -0,0 +1,511 @@
...
+          </tr>
+        </thead>
+        <tbody>
+          
+          <tr>
+            <td>1</td>
+            <td>9</td>
+            <td style="color: #22c55e">8</td>
+            <td style="color: #22c55e">0</td>
+            <td style="color: #22c55e">0</td>
```
Comment (stephane-segning) [inline plans/artillery/reports/report-qwen25-huge-prefix-lmcache.html:242]: 9 requests, 8 success, no error, no timeouts... I don't buy that math

File: plans/artillery/reports/report-qwen25-huge-prefix-lmcache.html [hunk 5]
```diff
@@ -0,0 +1,511 @@
...
+        </div>
+        <div class="percentile-item">
+          <div class="percentile-label">Max</div>
+          <div class="percentile-value">7,987 ms</div>
+        </div>
+      </div>
+    </div>
+    
+    
+    <div class="section">
```
Comment (stephane-segning) [inline plans/artillery/reports/report-qwen25-huge-prefix-lmcache.html:217]: Presented in a table, it looks like this:

Image

and it sure is not clean. Some numbers doesn't sum up.

**Review by stephane-segning — COMMENTED** [review]
Very nice adventure here. It was fun reading. I have some points tho

File: plans/artillery/justfile [hunk 6]
```diff
@@ -246,6 +246,193 @@ errors:
...
+# Generate the eviction-busters.csv file (~20 unique ~8K-token prompts)
+qwen25-gen-busters:
+    @echo "=== Generating eviction-busters.csv ==="
+    @cd self-hosted-model && node generate-eviction-busters.js
+
+# Run eviction (mixed buster + prefix-reuse) load test against current deploy
+qwen25-eviction: check-deps reports-dir qwen25-gen-busters
+    @test -n "$$VLLM_LOCAL_API_KEY" || { echo "❌ VLLM_LOCAL_API_KEY not set. Run: export VLLM_LOCAL_API_KEY=$$(just qwen25-api-key)"; exit 1; }
+    @echo "=== Approach B: Eviction Load Test ==="
+    artillery run self-hosted-model/artillery-qwen25-eviction.yml -o reports/report-qwen25-eviction.json
```
Comment (stephane-segning) [inline plans/artillery/justfile]: **This recipe points at a file that doesn't exist in the repo.** `artillery-qwen25-eviction.yml` is not in this PR nor on `main` (I checked both) — so `qwen25-eviction` (and `qwen25-approachb-full`, which chains into it) fails at runtime. Meanwhile the one artillery config this PR *does* add, `artillery-qwen25-huge-prefix.yml`, has **no recipe at all**. These look swapped: point this recipe (or a new `qwen25-huge-prefix` target) at the config the PR actually ships.

File: plans/artillery/justfile [hunk 7]
```diff
@@ -246,6 +246,193 @@ errors:
...
+    @just qwen25-prefix-reuse
+    @just qwen25-cold-unique
+    @just qwen25-kill-port-forward
+    @echo "=== Both tests complete ==="
+
+# Deploy baseline (no LMCache)
+qwen25-deploy-baseline:
+    @echo "=== Deploying baseline (no LMCache) ==="
+    helm upgrade --install model-serving-qwen25-3b-awq ./charts/model-serving-qwen25-3b-awq \
+      -n custom-ai-ops -f charts/model-serving-qwen25-3b-awq/local-overrides.yaml
```
Comment (stephane-segning) [inline plans/artillery/justfile]: **These deploy recipes reference override files that aren't in this PR or on `main`.** `local-overrides.yaml` (here) and `local-overrides-lmcache.yaml` (next recipe) don't exist — only `local-overrides-approachb.yaml` / `local-overrides-approachb-lmcache.yaml` are added by this PR. So `qwen25-deploy-baseline` / `qwen25-deploy-lmcache` (and `qwen25-full-ab`, `qwen25-run-both`, `qwen25-compare` that chain them) can't run standalone. They presumably depend on an unmerged sibling "Approach A" branch — either bring those files in, or drop/guard the recipes that need them so this PR is self-consistent.

File: plans/artillery/self-hosted-model/generate-huge-prefix.js [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+ *
+ * Run from plans/artillery/self-hosted-model/:
+ *   node generate-huge-prefix.js
+ *
+ * Output: huge-prefix.csv (columns: huge_prefix)
+ */
+const fs = require('fs');
+const path = require('path');
+
+const docPath = path.resolve(__dirname, '../../../docs/lmcache+vllm.md');
```
Comment (stephane-segning) [inline plans/artillery/self-hosted-model/generate-huge-prefix.js:32]: **Non-reproducible: this source file doesn't exist in the repo.** `generate-huge-prefix.js` reads `../../../docs/lmcache+vllm.md`, but `docs/lmcache+vllm.md` is not present on `main` and is not added by this PR — so the generator throws `ENOENT` for anyone re-running it. The committed `huge-prefix.csv` works, but the PR's stated goal ("reproducible CSV generation") isn't met for this generator. Either commit the source doc or vendor the prefix text into the script.

File: plans/artillery/self-hosted-model/artillery-qwen25-huge-prefix.yml [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+# APC is ENABLED on both variants. The override files are identical to v3
+# (approachb / approachb-lmcache) — only the Artillery payload differs.
+#
+# Payloads:
+#   self-hosted-model/huge-prefix.csv      — single-row full-md system prompt
+#   self-hosted-model/prefix-reuse-questions.csv — 15 rotating questions
+#   self-hosted-model/eviction-busters.csv — 20 unique ~2K-token busters
+#
+# Run from plans/artillery/ :
+#   just qwen25-huge-prefix
```
Comment (stephane-segning) [inline plans/artillery/self-hosted-model/artillery-qwen25-huge-prefix.yml:24]: **This run hint points at a recipe that doesn't exist.** The justfile in this PR adds no `qwen25-huge-prefix` target, so `just qwen25-huge-prefix` fails — this config can only be run via the raw `artillery run …` form below. Add the recipe (and wire it into `qwen25-approachb-full` in place of the missing `qwen25-eviction` config).

File: charts/model-serving-qwen25-3b-awq/local-overrides-approachb-lmcache.yaml [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+#
+# Experiment B keeps vLLM's APC ENABLED on both variants and forces its
+# eviction under a tight GPU KV budget + high concurrency. When APC misses
+# (evicted prefix), LMCache's CPU RAM tier (L1) retrieves the cached KV and
+# injects it into the GPU — skipping prefill. This mirrors real production:
+# GPU prefix cache misses under load, LMCache CPU tier saves the day.
+#
+# Key differences vs Experiment 0 (APC + LMCache, normal VRAM):
+#   - --gpu-memory-utilization=0.55  (was 0.90) → shrinks KV budget
+#   - --max-num-seqs=4               (was 2)   → more concurrent pressure
```
Comment (stephane-segning) [inline charts/model-serving-qwen25-3b-awq/local-overrides-approachb-lmcache.yaml]: **Contradictory comment.** Lines 11–12 list both `--max-num-seqs=4 (was 2)` and `--max-num-seqs=8 (was 2)` as the change; the actual arg (line 61) is `--max-num-seqs=8`. Drop the stale `=4` line.

File: charts/model-serving-qwen25-3b-awq/local-overrides-approachb-lmcache.yaml [hunk 11]
```diff
@@ -0,0 +1,134 @@
...
+  accessMode: ReadWriteOnce
+  size: 10Gi
+
+modelServing:
+  controllers:
+    main:
+      containers:
+        model:
+          image:
+            repository: lmcache/vllm-openai
```
Comment (stephane-segning) [inline charts/model-serving-qwen25-3b-awq/local-overrides-approachb-lmcache.yaml:56]: **A/B confounder: baseline and treatment run different vLLM builds.** Treatment is `lmcache/vllm-openai:v0.5.1`; the baseline (`local-overrides-approachb.yaml`) is `vllm/vllm-openai:v0.24.0`. That's unavoidable given LMCache ships its own vLLM image, but it means the two arms differ in more than "LMCache on/off," so some of the measured TTFT delta could be the engine build rather than LMCache. Worth an explicit caveat in the experiment report so the numbers aren't over-attributed to LMCache.

File: plans/artillery/self-hosted-model/generate-eviction-busters.js [hunk 12]
```diff
@@ -0,0 +1,103 @@
+#!/usr/bin/env node
+/**
+ * Generates eviction-busters.csv — ~20 unique large system prompts (~8000 tokens
```
Comment (stephane-segning) [inline plans/artillery/self-hosted-model/generate-eviction-busters.js]: **Docstring token count contradicts the generator.** Header says "~8000 tokens each" / "~32,000 chars → ~8000 tokens", but `generatePrompt` builds "17 paragraphs × ~120 tokens ≈ ~2000 tokens" and the final `console.log` prints "~2000" (which matches the artillery config's "~2K-token busters"). Fix the header to ~2000 so the reproducibility docs are accurate.

**Review by Koufan-De-King — APPROVED** [review]
